src/App.tsx

```diff
@@ -0,0 +1,798 @@
+    canv.setShowTrackPath(showTrackPath);
+    canv.setHideValuesOutOfRange(hideValuesOutOfRange);
+    canv.setColorRamp(colorRamp);
+    canv.setColorMapRangeMin(colorRampMin);
+    canv.setColorMapRangeMax(colorRampMax);
+    canv.setSelectedTrack(selectedTrack);
+    await canv.render();
+
+    // update current time in plot
+    plot?.setTime(currentFrame);
+
+    if (!timeControls?.isPlaying() && !recordingControls?.isRecording()) {
+      // Do not update URL while playback is happening for performance + UX reasons
+      updateUrl();
+    }
+  }, [
+    dataset,
+    datasetName,
+    featureName,
+    currentFrame,
+    selectedTrack,
+    hideValuesOutOfRange,
+    showTrackPath,
+    colorRamp,
+    colorRampMin,
+    colorRampMax,
+    timeControls?.isPlaying(), // updates URL when timeControls stops
+    updateUrl,
+  ]);
+
+  // Call drawLoop whenever its dependencies change.
+  useMemo(() => {
+    drawLoop();
+  }, [drawLoop]);
+
+  const setFrame = useCallback(
+    async (frame: number) => {
+      await canv.setFrame(frame);
+      setCurrentFrame(frame);
+      setFrameInput(frame);
+    },
+    [drawLoop, canv]
+  );
+
+  const findTrack = useCallback(
+    async (trackId: number, seekToFrame: boolean = true): Promise<void> => {
+      const newTrack = dataset!.buildTrack(trackId);
+
+      if (newTrack.length() < 1) {
+        // Check track validity
+        return;
+      }
+      setSelectedTrack(newTrack);
+      if (seekToFrame) {
+        setFrame(newTrack.times[0]);
+        plot?.plot(newTrack, featureName, newTrack.times[0]);
+      } else {
+        plot?.plot(newTrack, featureName, currentFrame);
+      }
+      setFindTrackInput("" + trackId);
+      updateUrl();
+    },
+    [canv, plot, dataset, featureName, currentFrame]
+  );
+
+  /**
+   * Update the canvas dimensions based on the current window size.
+   */
+  const setSize = (): void => canv.setSize(Math.min(window.innerWidth, 730), Math.min(window.innerHeight, 500));
+
+  // INITIAL SETUP  ////////////////////////////////////////////////////////////////
+
+  // Use a memoized value here and only retrieve parameters once, because the URL can be updated
+  // by state updates and lose information (like the track, feature, time, etc.) that isn't
+  // accessed in the first render.
+  const initialUrlParams = useMemo(urlUtils.loadParamsFromUrl, []);
+
+  // Load database and collections data from the URL.
+  // This is memoized so that it only runs one time on startup.
+  useMemo(async () => {
+    setSize();
+
+    let _collection = initialUrlParams.collection;
+    const _datasetName = initialUrlParams.dataset;
+    let _collectionData;
+    // Load single dataset instead of collection
+    if (_datasetName && urlUtils.isUrl(_datasetName)) {
+      await replaceDataset(_datasetName);
+      setIsInitialDatasetLoaded(true);
+      return;
+    }
+
+    // Load collection data.
+    _collection = _collection || urlUtils.DEFAULT_COLLECTION_PATH;
```

Comment:
I think you have this as a TODO, but pulling out all the url handling and data initialization out of a render function would make this easier to follow. but I realize this is a first pass at conversion

Reply:
You're right, that makes sense. There's a refactor I wanted to do related to the way collections are currently being handled that I think this change would slot into neatly. I'll add a TODO in the code itself about this!

Reply:
There's also something about factoring the bodies of many of the functions out into other modules that would also make them potentially more unit-testable outside of React?  Just a vague thought for future work.